app/services/analytics.rb

```diff
@@ -7,9 +7,11 @@ def initialize(user, request)
   def track_event(event, attributes = { user_id: uuid })
     attributes[:user_id] = uuid unless attributes.key?(:user_id)
 
-    Rails.logger.info("#{event}: #{attributes}")
+    consolidated_attributes = attributes.merge!(request_attributes)
```

Comment:
For performance reasons. `merge!` is faster than `merge` because it doesn't create a new hash. The attributes input is no longer used after the merge, so it's safe to mutate IMO. We always want to send the input attributes plus a set of global attributes.